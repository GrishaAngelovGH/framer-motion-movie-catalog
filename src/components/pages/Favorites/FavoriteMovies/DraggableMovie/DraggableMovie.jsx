import { motion, useMotionValue, useTransform } from "framer-motion"

import persistentMovieCatalog from "persistent/persistentMovieCatalog"

const DraggableMovie = ({ movie: { id, image }, updateFavoriteMovies, onBackgroundChange }) => {
  // The component calculates the current background color 
  // when starting to drag to the right and applies it dynamically.

  const x = useMotionValue(0)

  const xInput = [0, 100]

  const background = useTransform(x, xInput, [
    "linear-gradient(180deg, #1e40af 0%, rgb(30, 64, 175) 100%)",
    "linear-gradient(180deg, #ff008c 0%, rgb(211, 9, 225) 100%)"
  ])

  return (
    <motion.img
      style={{ x }}
      drag="x"
      src={`/images/${image}`}
      className="m-8 flex justify-center items-center rounded-md w-[200px] h-[350px] cursor-pointer"

      // Disable drag to left, the right degree of movement is set to default 
      dragElastic={{ left: 0, right: 0.5 }}

      // Applies constraints on the permitted draggable area.
      // It can accept an object of optional top, left, right, and bottom values,
      // measured in pixels. This will define a distance the named edge of the draggable component.
      dragConstraints={{ left: 0, right: 0 }}

      onDragStart={(e) => {
        onBackgroundChange(background)
      }}

      onDragEnd={(e) => {
        if (e.clientX < 400) {
          return
        }

        persistentMovieCatalog.removeAsFavorite(id)
        onBackgroundChange("linear-gradient(180deg, #1e40af 0%, rgb(30, 64, 175) 100%)")
        updateFavoriteMovies()
      }}
    >
    </motion.img>
  )
}

export default DraggableMovie