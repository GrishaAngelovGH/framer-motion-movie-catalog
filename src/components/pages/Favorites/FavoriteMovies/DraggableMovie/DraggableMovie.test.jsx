import { render } from "@testing-library/react"

import DraggableMovie from "./DraggableMovie"

test("should render DraggableMovie component", () => {
  const view = render(
    <DraggableMovie movie={{ image: "image-1.jpg" }} />
  )

  expect(view).toMatchSnapshot()
})