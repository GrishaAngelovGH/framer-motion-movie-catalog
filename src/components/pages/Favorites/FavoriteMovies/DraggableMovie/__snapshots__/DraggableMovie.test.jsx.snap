// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render DraggableMovie component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <img
        class="m-8 flex justify-center items-center rounded-md w-[200px] h-[350px] cursor-pointer"
        draggable="false"
        src="/images/image-1.jpg"
        style="transform: none; user-select: none;"
      />
    </div>
  </body>,
  "container": <div>
    <img
      class="m-8 flex justify-center items-center rounded-md w-[200px] h-[350px] cursor-pointer"
      draggable="false"
      src="/images/image-1.jpg"
      style="transform: none; user-select: none;"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
